File: frontend/types/clientCompanies.ts
export type ClientCompanyStatus = "active" | "inactive";

export type ClientCompany = {
  id: string;
  companyName: string;
  taxCode: string | null;
  address: string | null;
  contactPerson: string | null;
  phoneNumber: string | null;
  email: string | null;
  description: string | null;
  isActive: boolean;
  createdByUserId: string;
  createdAt: string;
  updatedAt: string;
};

export type ClientCompanyApiResponse = {
  id: string;
  company_name: string;
  tax_code: string | null;
  address: string | null;
  contact_person: string | null;
  phone_number: string | null;
  email: string | null;
  description: string | null;
  is_active: boolean;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
};

export type ClientCompanyListApiResponse = {
  records: ClientCompanyApiResponse[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type ClientCompanyListResponse = {
  records: ClientCompany[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
};

export type ClientCompanyFormValues = {
  companyName: string;
  taxCode: string;
  address: string;
  contactPerson: string;
  phoneNumber: string;
  email: string;
  description: string;
};

export type ClientCompanyListQuery = {
  search?: string;
  isActive?: boolean;
  page: number;
  pageSize: number;
  sortBy: "company_name" | "tax_code" | "created_at" | "updated_at" | "is_active";
  sortOrder: "asc" | "desc";
};

File: backend/app/schemas/client_companies.py
import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class ClientCompanySortBy(str, Enum):
    COMPANY_NAME = "company_name"
    TAX_CODE = "tax_code"
    CREATED_AT = "created_at"
    UPDATED_AT = "updated_at"
    IS_ACTIVE = "is_active"


class SortOrder(str, Enum):
    ASC = "asc"
    DESC = "desc"


class ClientCompanyBase(BaseModel):
    company_name: str = Field(min_length=1, max_length=255)
    tax_code: str | None = Field(default=None, max_length=100)
    address: str | None = None
    contact_person: str | None = Field(default=None, max_length=255)
    phone_number: str | None = Field(default=None, max_length=50)
    email: EmailStr | None = None
    description: str | None = None

    @field_validator(
        "company_name",
        "tax_code",
        "address",
        "contact_person",
        "phone_number",
        "description",
        mode="before",
    )
    @classmethod
    def normalize_optional_text(cls, raw_value: str | None) -> str | None:
        if raw_value is None:
            return None

        normalized_value = raw_value.strip()
        return normalized_value or None


class ClientCompanyCreate(ClientCompanyBase):
    pass


class ClientCompanyUpdate(BaseModel):
    company_name: str | None = Field(default=None, min_length=1, max_length=255)
    tax_code: str | None = Field(default=None, max_length=100)
    address: str | None = None
    contact_person: str | None = Field(default=None, max_length=255)
    phone_number: str | None = Field(default=None, max_length=50)
    email: EmailStr | None = None
    description: str | None = None

    @field_validator(
        "company_name",
        "tax_code",
        "address",
        "contact_person",
        "phone_number",
        "description",
        mode="before",
    )
    @classmethod
    def normalize_optional_text(cls, raw_value: str | None) -> str | None:
        if raw_value is None:
            return None

        normalized_value = raw_value.strip()
        return normalized_value or None


class ClientCompanyStatusUpdate(BaseModel):
    is_active: bool


class ClientCompanyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    company_name: str
    tax_code: str | None
    address: str | None
    contact_person: str | None
    phone_number: str | None
    email: EmailStr | None
    description: str | None
    is_active: bool
    created_by_user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ClientCompanyListResponse(BaseModel):
    records: list[ClientCompanyRead]
    total: int
    page: int
    page_size: int
    total_pages: int
